src/stdlib_32_bit_nmhashes.fypp

```diff
@@ -0,0 +1,797 @@
+!! LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+!! CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+!! SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+!! INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+!! CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+!! ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+!! POSSIBILITY OF SUCH DAMAGE.
+!!------------------------------------------------------------------------------
+
+#! Integer kinds to be considered during templating
+#:set INT_KINDS = ["int16", "int32", "int64"]
+
+submodule(stdlib_32_bit_hash_codes) stdlib_32_bit_nmhashes
+
+    implicit none
+
+! Primes from XXH
+    integer(int32), parameter :: nmh_prime32_1 = int( Z'9E3779B1', int32 )
+    integer(int32), parameter :: nmh_prime32_2 = int( Z'85EBCA77', int32 )
+    integer(int32), parameter :: nmh_prime32_3 = int( Z'C2B2AE3D', int32 )
+    integer(int32), parameter :: nmh_prime32_4 = int( Z'27D4EB2F', int32 )
+
+    integer(int32), parameter :: nmh_m1 = int(z'F0D9649B', int32 )
+    integer(int32), parameter :: nmh_m2 = int(z'29A7935D', int32 )
+    integer(int32), parameter :: nmh_m3 = int(z'55D35831', int32 )
+
+    integer(int32), parameter :: nmh_m1_v(0:31) = nmh_m1
+    integer(int32), parameter :: nmh_m2_v(0:31) = nmh_m2
+    integer(int32), parameter :: nmh_m3_v(0:31) = nmh_m3
+
+    integer(int16), parameter :: nmh_m3_16(2) = transfer( nmh_m3, 0_int16, 2 )
+
+    logical, parameter :: nmh_short32_without_seed2=.false.
+    logical, parameter :: nmh_short32_with_seed2=.true.
+
+    integer, parameter :: init_size = 32
+
+! Pseudorandom secrets taken directly from FARSH.
+    integer(int32), parameter :: nmh_acc_init(0:init_size-1) = [ &
+        int( z'B8FE6C39', int32 ), int( z'23A44BBE', int32 ), &
+        int( z'7C01812C', int32 ), int( z'F721AD1C', int32 ), &
+        int( z'DED46DE9', int32 ), int( z'839097DB', int32 ), &
+        int( z'7240A4A4', int32 ), int( z'B7B3671F', int32 ), &
+        int( z'CB79E64E', int32 ), int( z'CCC0E578', int32 ), &
+        int( z'825AD07D', int32 ), int( z'CCFF7221', int32 ), &
+        int( z'B8084674', int32 ), int( z'F743248E', int32 ), &
+        int( z'E03590E6', int32 ), int( z'813A264C', int32 ), &
+
+        int( z'3C2852BB', int32 ), int( z'91C300CB', int32 ), &
+        int( z'88D0658B', int32 ), int( z'1B532EA3', int32 ), &
+        int( z'71644897', int32 ), int( z'A20DF94E', int32 ), &
+        int( z'3819EF46', int32 ), int( z'A9DEACD8', int32 ), &
+        int( z'A8FA763F', int32 ), int( z'E39C343F', int32 ), &
+        int( z'F9DCBBC7', int32 ), int( z'C70B4F1D', int32 ), &
+        int( z'8A51E04B', int32 ), int( z'CDB45931', int32 ), &
+        int( z'C89F7EC9', int32 ), int( z'D9787364', int32 ) ]
+
+contains
+
+    pure function nmh_readle32( p ) result( v )
```

Comment:
Can it be elemental?

Reply:
In principle yes, but it is a private procedure and never used except on a scalar.

Reply:
In thinking further, no, its sole argument is a vector.